src/ClientTransactionTask.cc

```diff
@@ -0,0 +1,562 @@
+/* Copyright (c) 2015 Stanford University
+ *
+ * Permission to use, copy, modify, and distribute this software for any
+ * purpose with or without fee is hereby granted, provided that the above
+ * copyright notice and this permission notice appear in all copies.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR(S) DISCLAIM ALL WARRANTIES
+ * WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
+ * MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL AUTHORS BE LIABLE FOR
+ * ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+ * WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
+ * ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+ * OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+ */
+
+#include "ClientTransactionTask.h"
+#include "RamCloud.h"
+
+namespace RAMCloud {
+
+/**
+ * Constructor for a transaction task.
+ *
+ * \param ramcloud
+ *      Overall information about the calling client.
+ */
+ClientTransactionTask::ClientTransactionTask(RamCloud* ramcloud)
+    : ramcloud(ramcloud)
+    , participantCount(0)
+    , participantList()
+    , state(INIT)
+    , status(STATUS_OK)
+    , decision(WireFormat::TxDecision::COMMIT)
+    , lease()
+    , txId(0)
+    , prepareRpcs()
+    , decisionRpcs()
+    , commitCache()
+    , nextCacheEntry()
+{
+}
+
+/**
+ * Find and return the cache entry identified by the given key.
+ *
+ * \param key
+ *      Key of the object contained in the cache entry that should be returned.
+ * \return
+ *      Returns a pointer to the cache entry if found.  Returns NULL otherwise.
+ *      Pointer is invalid once the commitCache is modified.
+ */
+ClientTransactionTask::CacheEntry*
+ClientTransactionTask::findCacheEntry(Key& key)
+{
+    CacheKey cacheKey = {key.getTableId(), key.getHash()};
+    CommitCacheMap::iterator it = commitCache.lower_bound(cacheKey);
+    CacheEntry* entry = NULL;
+
+    while (it != commitCache.end()) {
+        if (cacheKey < it->first) {
+            break;
+        } else if (it->second.objectBuf) {
+            Key otherKey(it->first.tableId,
+                         it->second.objectBuf->getKey(),
+                         it->second.objectBuf->getKeyLength());
+            if (key == otherKey) {
+                entry = &it->second;
+                break;
+            }
+        }
+        it++;
+    }
+    return entry;
+}
+
+/**
+ * Return the transaction commit decision if a decision has been reached.
+ * Otherwise, INVALID will be returned.
+ */
+WireFormat::TxDecision::Decision
+ClientTransactionTask::getDecision()
+{
+    if (state != DECISION && state != DONE)
+        return WireFormat::TxDecision::INVALID;
+    return decision;
+}
+
+/**
+ * Inserts a new cache entry with the provided key and value.  Other members
+ * of the cache entry are left to their default values.  This method must not
+ * be called once the transaction has started committing.
```

Comment:
Must the caller ensure that no such entry already exists? The code below appears to assume that.